Added an IPv4/IPv6 toggle to TelegramClient constructor

solarin opened the pull request.

knocte commented:
@solarin now you forgot to include the changes to the .csproj file :)

And please configure your editor to use 4 spaces, not tabs.

solarin commented:
> @solarin now you forgot to include the changes to the .csproj file :)
> And please configure your editor to use 4 spaces, not tabs.

i'm getting crazy

solarin commented:
also added some methods that i needed for my project


+ GetFullChat to get the info of a chat 
+ GetAllChats to get all available user's chats
+ MakeAuthBotAsync to authenticate a bot instead of a user

knocte commented:
> also added some methods that i needed for my project

Please don't include those here, open a different PullRequest for that.

knocte commented on TLSharp.Core/DataCenterIPVersion.cs:
```diff
@@ -0,0 +1,30 @@
+        /// </summary>
+        PreferIPv4 = 3,
+        /// <summary>
+        /// Connection to IPv6 addresses is preferred to IPv6 addresses
```
typo here ^

solarin replied:
strange, i changed it O.o

knocte replied:
yes, this one was fixed, but I still see the text “Prefers IPv6 addresses if any is passed by Telegram” describing “Default”

solarin replied:
it's what actually does

knocte replied:
No, the text “Prefers IPv6 addresses if any is passed by Telegram” should describe "PreferIPv6", not "Default".

knocte replied:
remember Default doesn't do any filtering, just chooses the first datacenter's IP given by telegram servers

solarin replied:
it doesn't do any filtering, but it's a side effect of the algorithm to pick an ipv6 if any is passed just because it will be at the end of the list.

anyways, because the default behavious might change in the future, we can remove this comment.

knocte replied:
we can fix the side-effect, don't do any OrderBy if .Default is chosen please.

knocte commented on TLSharp.Core/DataCenterIPVersion.cs:
```diff
@@ -0,0 +1,30 @@
+	public enum DataCenterIPVersion
+    {
+        /// <summary>
+        /// Prefers IPv6 addresses if any is passed by Telegram
```
bad copy-paste here ^

solarin replied:
fixed

knocte replied:
this one is not fixed

knocte commented on TLSharp.Core/DataCenterIPVersion.cs:
```diff
@@ -0,0 +1,30 @@
+    /// When the Telegram server responds with a set of addresses to connect to, DataCenterIPVersion indicates a preference 
+    /// for how to choose the IP address to connect to 
+    /// </summary>
+	public enum DataCenterIPVersion
```
this line still has a tab instead of 4 spaces

knocte commented on TLSharp.Core/TelegramClient.cs:
```diff
@@ -51,9 +64,10 @@ public Session Session
 
             session = Session.TryLoadOrCreateNew(store, sessionUserId);
-            transport = new TcpTransport (session.DataCenter.Address, session.DataCenter.Port, this.handler);
+            transport = new TcpTransport(session.DataCenter.Address, session.DataCenter.Port, this.handler);
```
no need to change this line above ^

solarin replied:
when i swapped the editor to replace tabs with the spaces, i had to autoformat and this was auto formatted as well.

knocte replied:
well then revert the autoformat, you're breaking the git-blame for the lines that don't need changing

knocte commented on TLSharp.Core/TelegramClient.cs:
```diff
@@ -123,15 +151,15 @@ private async Task RequestWithDcMigration(TLMethod request, CancellationToken to
 
             var completed = false;
-            while(!completed)
+            while (!completed)
```
same here ^

solarin replied:
same as above

knocte commented on TLSharp.Core/TelegramClient.cs:
```diff
@@ -123,15 +151,15 @@ private async Task RequestWithDcMigration(TLMethod request, CancellationToken to
                     completed = true;
                 }
-                catch(DataCenterMigrationException e)
+                catch (DataCenterMigrationException e)
```
etc...

solarin replied:
same as above

solarin commented:
> also added some methods that i needed for my project
> 
> Please don't include those here, open a different PullRequest for that.

ok i deleted these and created a new PR

solarin commented:
i think i have changed all the spaces back to the originals

knocte commented:
> i think i have changed all the spaces back to the originals

You missed many, sorry, click here and you will see them all: https://github.com/sochix/TLSharp/pull/925/files

knocte commented on TLSharp.Core/DataCenterIPVersion.cs:
```diff
@@ -0,0 +1,30 @@
+        /// <summary>
+        /// Prefers IPv6 addresses if any is passed by Telegram
+        /// </summary>
+        Default = 0,
```
why do we need this enum be int-based? there's no need

solarin replied:
there is no need.

knocte replied:
cool let's remove the assignments to integers then

knocte commented:
Thanks for your patience. I will merge this tonite or tomorrow.

solarin commented:
i propose to rename DataCenterIPVersion to DataCenterIPVersions
and also move this file under namespace TLSharp.Core.Types

because we need other enums and other types that we need logically, although they have a different name or type at protocol-level

One example is ParticipantTypes (Recents, 
        Restricted,
        Admins,
        Bots,
        Search,
        Contacts,
        Kicked,
        Banned)

which at protcol level exists in different ways.

SALEH-SOFTWARE commented:
var client = new TelegramClient(apiId, apiHash);
   await client.ConnectAsync(); 

this code not have dependency !! How Can Now Connect To Telegram? And Please Put Poll Method For Vote Channel

knocte commented:
> this code not have dependency !

what dependency? what are you talking about?
